### PR TITLE
KAFKA-12336 Custom stream naming does not work while calling stream[K…

### PR DESCRIPTION
**Pull request by g1geordie**
Custom stream naming does not work while calling stream[K, V](topicPattern: Pattern)

### CODE REVIEW

**Comment by g1geordie** [comment]
@mjsax Hello~ 
Can you help me to take a look?

**Comment by mjsax** [comment]
Call for review @bbejeck

@g1geordie there seem to be failing tests (`SuppressTopologyTest` and `RepartitionOptimizingTest `) relevant to this change:

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/NamedInternal.java [hunk 1]
```diff
@@ -57,40 +59,28 @@ public String name() {
...
-            return suffixed;
-        } else {
-            return provider.newProcessorName(prefix);
-        }
+        return suffixWithOrElseGet(suffix, defaultName);
     }
 
     String orElseGenerateWithPrefix(final InternalNameProvider provider, final String prefix) {
         // We actually do not need to generate processor names for operation if a name is specified.
         // But before returning, we still need to burn index for the operation to keep topology backward compatibility.
```
Comment (g1geordie) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/NamedInternal.java]: May I ask a question?
 
Is Backward compatibility still in the current version?
I see the hint in other test

<pre><code>
SuppressTopologyTest.shouldApplyNameToSuppressionNode {
        ...
        .suppress(untilTimeLimit(Duration.ofSeconds(1), unbounded()).withName("asdf"))
        ....
       // <del>without</del>(with) the name,  <b>the suppression node does not increment the topology index </b>
        assertThat(namedNodeTopology, is(NAMED_INTERMEDIATE_TOPOLOGY));
}

Comment (g1geordie) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/NamedInternal.java]: The NAMED_INTERMEDIATE_TOPOLOGY
<pre><code>
Topologies:
   Sub-topology: 0
    <b>Source: KSTREAM-SOURCE-0000000000 (topics: [input])</b>
      --> KSTREAM-AGGREGATE-0000000002
    <b>Processor: KSTREAM-AGGREGATE-0000000002 (stores: [KSTREAM-AGGREGATE-STATE-STORE-0000000001])</b>
      --> asdf
      <-- KSTREAM-SOURCE-0000000000
    <b>Processor: asdf (stores: [asdf-store])</b>
      --> KTABLE-TOSTREAM-0000000003
      <-- KSTREAM-AGGREGATE-0000000002
    <b>Processor: KTABLE-TOSTREAM-0000000003 (stores: [])</b>
      --> KSTREAM-SINK-0000000004
      <-- asdf
    <b>Sink: KSTREAM-SINK-0000000004 (topic: output)</b>
      <-- KTABLE-TOSTREAM-0000000003
</code></pre>

**Comment by bbejeck** [comment]
@g1geordie, thanks for the PR and sorry for the delay in the review.  

Can you rebase this PR to resolve the conflicts?  I plan on reviewing this in the next day or so.  Also, there seem to be some failing tests that are related.

Thanks!

**Comment by g1geordie** [comment]
@bbejeck 
Thanks for your response !!

I see the compatibility claim in NamedInternal.
```
String orElseGenerateWithPrefix(final InternalNameProvider provider, final String prefix) {
        // We actually do not need to generate processor names for operation if a name is specified.
        // But before returning, we still need to burn index for the operation to keep topology backward compatibility.
        ...
}
```
The index is not correct  in SuppressTopologyTest.shouldApplyNameToSuppressionNode.
	  
```
without  `.suppress(... .withName("asdf"))` name option.
	
Topologies:
   Sub-topology: 0
    Source: KSTREAM-SOURCE-0000000000 (topics: [input])
      --> KSTREAM-AGGREGATE-0000000002
    Processor: KSTREAM-AGGREGATE-0000000002 (stores: [KSTREAM-AGGREGATE-STATE-STORE-0000000001])
      --> KTABLE-SUPPRESS-0000000003
      <-- KSTREAM-SOURCE-0000000000
    Processor: KTABLE-SUPPRESS-0000000003 (stores: [KTABLE-SUPPRESS-STATE-STORE-0000000004])
      --> KTABLE-TOSTREAM-0000000005
      <-- KSTREAM-AGGREGATE-0000000002
    Processor: KTABLE-TOSTREAM-0000000005 (stores: [])
      --> KSTREAM-SINK-0000000006
      <-- KTABLE-SUPPRESS-0000000003
    Sink: KSTREAM-SINK-0000000006 (topic: output)
      <-- KTABLE-TOSTREAM-0000000005
```
So I change expected index to below .
```
Topologies:
   Sub-topology: 0
    Source: KSTREAM-SOURCE-0000000000 (topics: [input])
      --> KSTREAM-AGGREGATE-0000000002
    Processor: KSTREAM-AGGREGATE-0000000002 (stores: [KSTREAM-AGGREGATE-STATE-STORE-0000000001])
      --> asdf
      <-- KSTREAM-SOURCE-0000000000
    Processor: asdf (stores: [asdf-store])
      --> KTABLE-TOSTREAM-0000000005
      <-- KSTREAM-AGGREGATE-0000000002
    Processor: KTABLE-TOSTREAM-0000000005 (stores: [])
      --> KSTREAM-SINK-0000000006
      <-- asdf
    Sink: KSTREAM-SINK-0000000006 (topic: output)
      <-- KTABLE-TOSTREAM-0000000005  
```

**Comment by bbejeck** [comment]
@g1geordie apologies for letting this go for so long.  I'm looking now and let's see if we can't get this in over the next few days.

**Review by bbejeck — CHANGES_REQUESTED** [review]
Thanks for the patch @g1geordie!  

Overall this looks good, but I think some of the changes here go beyond the scope of the ticket. 

I say this because the issue is about using a provided name when using the  `builder.source(Pattern..., Consumed.with(myName))`  method.

But there are modifications to the topology names in  the `SuppessTopologyTest` and the `RepartitionOptimizingTest` to get the tests to pass, but neither test uses a source with a `Pattern`.  I wouldn't expect the topology names to shift as a side effect in this case.

So here's what I suggest:

1.  For this PR just put in the change to use a provided name with a `StreamBuilder.source(Pattern, Consumed.with("someName"))` along with a unit test.
2. Submit a follow-up Jira ticket and PR and with the other changes you've made.  Please note that making any changes which causes the existing topology names to shift is something we try to avoid as it would cause problems for existing streams users simply by upgrading.

WDYT?

Thanks again for your contribution!

**Comment by g1geordie** [comment]
@bbejeck  Thanks for the review.  
You are right !!!!
Please help me review again.

**Review by bbejeck — APPROVED** [review]
Thanks for the update @g1geordie.  LGTM, just waiting for tests to pass before merging.

**Comment by g1geordie** [comment]
@bbejeck 
Failed test also appear in other issues. I think it's unrelated.
```
org.apache.kafka.common.network.SslTransportLayerTest.[1] tlsProtocol=TLSv1.2, useInlinePem=false
kafka.server.RaftClusterTest.testCreateClusterAndCreateListDeleteTopic()
kafka.server.RaftClusterTest.testCreateClusterAndCreateAndManyTopicsWithManyPartitions()
```
Thank you.

**Comment by bbejeck** [comment]
```
Build / JDK 11 and Scala 2.13 / [1] tlsProtocol=TLSv1.2, useInlinePem=false – org.apache.kafka.common.network.SslTransportLayerTest
15s
Build / JDK 11 and Scala 2.13 / [2] tlsProtocol=TLSv1.2, useInlinePem=true – org.apache.kafka.common.network.SslTransportLayerTest
15s
Build / JDK 11 and Scala 2.13 / [3] tlsProtocol=TLSv1.3, useInlinePem=false – org.apache.kafka.common.network.SslTransportLayerTest
15s
Build / JDK 11 and Scala 2.13 / [1] tlsProtocol=TLSv1.2, useInlinePem=false – org.apache.kafka.common.network.SslTransportLayerTest
15s
Build / JDK 11 and Scala 2.13 / [2] tlsProtocol=TLSv1.2, useInlinePem=true – org.apache.kafka.common.network.SslTransportLayerTest
15s
Build / JDK 11 and Scala 2.13 / [3] tlsProtocol=TLSv1.3, useInlinePem=false – org.apache.kafka.common.network.SslTransportLayerTest
15s
Build / JDK 11 and Scala 2.13 / testCreateClusterAndCreateAndManyTopics() – kafka.server.RaftClusterTest
16s
Build / JDK 16 and Scala 2.13 / testCreateClusterAndCreateListDeleteTopic() – kafka.server.RaftClusterTest
23s
Build / JDK 8 and Scala 2.12 / testCreateClusterAndCreateAndManyTopicsWithManyPartitions() – kafka.server.RaftClusterTest
16s
```
Failures unrelated to this PR

**Comment by bbejeck** [comment]
merged #10190 into trunk

**Comment by bbejeck** [comment]
Thanks @g1geordie for the contribution!

**Comment by bbejeck** [comment]
cherry-picked to 2.8

**Pull request merged**